Refactor user_ldap group membership cache and add check-group command

come-nc opened the pull request:
* Resolves: #29727

## Summary

Refactor user_ldap group membership cache, to be able to easily check if a membership is known or not.
Add a command check-group which works like check-user, with the same --update option to force update from LDAP.

## TODO

- [x] Figure out if indexes are needed on the new table
- [x] Improve command output (maybe use event listeners?)
- [x] Make sure command ignores cache for `--update` (currently not the case)

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required https://github.com/nextcloud/documentation/pull/11079
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

come-nc committed:
Refactor user_ldap group membership to use flat DB

Move away from serialized arrays. Also use a QBMapper class for the new table.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Update autoloader mapping for user_ldap

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add group events to UpdateGroups

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Move UpdateGroups methods to a service

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add check-group command

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
```json
{
"level":3,
"message":"Table \"oc_ldap_group_membership\" has no primary key and therefor will not behave sane in clustered setups. This will throw an exception and not be installable in a future version of Nextcloud.",
}
```

- [ ] Add primary key

come-nc committed:
Add missing primary key for ldap_group_membership

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix errors in UpdateGroupsService.php

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Always empty cache before updating a group

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add output to check-group --update command

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
- [ ] It only works if using full dn for non-mapped groups, would be better if using name works as well (for user it’s complicated but for groups it might be simpler to do that)
- [ ] It does not seem to correctly fire the group created event when passed a full dn of a new group

come-nc committed:
Fix getKnownGroups return value

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix check-group command for new groups

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
The `GroupCreatedEvent` firing is not completely working, but since GroupDeletedEvent is not fired I’m letting that for a follow-up PR.

Altahrim commented on apps/user_ldap/lib/Command/CheckGroup.php:
```diff
@@ -0,0 +1,173 @@
+					$this->updateGroup($gid, $output, $wasMapped);
+				}
+				return 0;
+			} elseif ($wasMapped) {
```
Since you use `return`, we can avoid the `else`:
```suggestion
			}
			if ($wasMapped) {
```
Same for lines `109-111`

come-nc replied:
I find it clearer with the else.

Altahrim commented on apps/user_ldap/lib/Db/GroupMembershipMapper.php:
```diff
@@ -0,0 +1,77 @@
+			->from($this->getTableName())
+			->executeQuery();
+
+		$groups = array_map(fn ($row) => $row['groupid'], $result->fetchAll());
```
No need to bind `$this`:
```suggestion
		$groups = array_column($result->fetchAll(), 'groupid');
```

Or maybe this one with DBAL: 
```suggestion
		$groups = $result->fetchFirstColumn();
```

come-nc replied:
`fetchFirstColumn` does not exists on `OCP\DB\IResult`.
I’ll switch to array_column.

Altahrim commented on apps/user_ldap/lib/Migration/Version1190Date20230706134108.php:
```diff
@@ -0,0 +1,113 @@
+
+		if (!$schema->hasTable('ldap_group_membership')) {
+			$table = $schema->createTable('ldap_group_membership');
+			$table->addColumn('id', 'integer', [
```
If we only need `length = 4`, `SMALLINT` should be enough:
```suggestion
			$table->addColumn('id', Types:SMALLINT, [
```

come-nc replied:
Hum this was from a copy/paste I think, from what I understood all ids should be BIGINT. I’ll switch to bigint, there might be millions of users in millions of groups.

Altahrim replied:
Why keep `length = 4` then?

Altahrim commented on apps/user_ldap/lib/Service/UpdateGroupsService.php:
```diff
@@ -0,0 +1,188 @@
+			$this->logger->debug('service "updateGroups" – Dealing with {group}.', ['group' => $group]);
+			$groupMemberships = $this->groupMembershipMapper->findGroupMemberships($group);
+			$knownUsers = array_map(
+				fn (GroupMembership $groupMembership): string => $groupMembership->getUserid(),
```
```suggestion
				static fn (GroupMembership $groupMembership): string => $groupMembership->getUserid(),
```
I think we can avoid binding `$this `

come-nc committed:
[user_ldap] Small perf improvements suggestion from review

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Dispatch UserRemovedEvents for removed groups

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Removed GroupCreatedEvent/GroupDeletedEvent from UpdateGroupsService

This should be handled when mapping groups, not when registering their
 members. An empty group may still exist.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Note: I looked into using LazyUser for events, but the test expects the event to only fire for existing users, which makes sense I suppose, so I abandoned the idea.

come-nc committed:
Use BIGINT for new table id field

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix check-group --update for deleted groups

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add unique index for user_ldap group memberships

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Rebased and squashed
